sys/net/gnrc/netif/gnrc_netif.c

```diff
@@ -1454,86 +1463,183 @@ static void _process_receive_stats(gnrc_netif_t *netdev, gnrc_pktsnip_t *pkt)
+        netdev_t *dev = netif->dev;
+        dev->driver->get(dev, NETOPT_TX_RETRIES_NEEDED, &retries, sizeof(retries));
+        netstats_nb_update_tx(&netif->netif, NETSTATS_NB_SUCCESS, retries + 1);
+    }
+    /* send packet previously queued within netif due to the lower
+     * layer being busy.
+     * Further packets will be sent on later TX_COMPLETE or
+     * TX_MEDIUM_BUSY */
+    return _send_queued_pkt(netif);
 }
 
+
 /**
  * @brief   Process any pending events and wait for IPC messages
  *
- * This function will block until an IPC message is received. Events posted to
- * the event queue will be processed while waiting for messages.
+ * This function will block until an IPC message is received. ISR events
+ * will still be processed.
  *
- * @param[in]   netif   gnrc_netif instance to operate on
- * @param[out]  msg     pointer to message buffer to write the first received message
+ * @param[in]   netif       gnrc_netif instance to operate on
+ * @param[in]   dev         netdev corresponding to @p netif
+ * @param[out]  msg         pointer to message buffer to write the first received
+ *                          message
+ * @param[in]   tx_pkt      frame that is currently being send, or `NULL` if no transmission is
+ *                          currently ongoing
  *
- * @return >0 if msg contains a new message
+ * @return  Frame that is currently being send
+ * @retval  NULL            No frame is currently being send
  */
-static void _process_events_await_msg(gnrc_netif_t *netif, msg_t *msg)
+static gnrc_pktsnip_t * _process_events_await_msg(gnrc_netif_t *netif, netdev_t *dev, msg_t *msg,
+                                                  gnrc_pktsnip_t *tx_pkt)
 {
-    if (IS_USED(MODULE_GNRC_NETIF_EVENTS)) {
-        while (1) {
-            /* Using messages for external IPC, and events for internal events */
-
-            /* First drain the queues before blocking the thread */
-            /* Events will be handled before messages */
-            DEBUG("gnrc_netif: handling events\n");
-            event_t *evp;
-            /* We can not use event_loop() or event_wait() because then we would not
-             * wake up when a message arrives */
-            event_queue_t *evq = _get_evq(netif);
-            while ((evp = event_get(evq))) {
-                DEBUG("gnrc_netif: event %p\n", (void *)evp);
-                if (evp->handler) {
-                    evp->handler(evp);
-                }
+    DEBUG("gnrc_netif: waiting for incoming events or messages\n");
+    const thread_flags_t flags_mask = THREAD_FLAG_IRQ | THREAD_FLAG_RX_DONE | THREAD_FLAG_TX_DONE
+                                    | THREAD_FLAG_MSG_WAITING;
+    while (msg_try_receive(msg) == -1) {
+        thread_flags_t flags = thread_flags_wait_any(flags_mask);
+        if (flags & THREAD_FLAG_IRQ) {
+            dev->driver->isr(dev);
+        }
+        if (flags & THREAD_FLAG_RX_DONE) {
+            gnrc_pktsnip_t *pkt = netif->ops->recv(netif);
```

Comment:
Do you think it is reasonable to have something like this:
```C
if (flags & THREAD_FLAG_RX_DONE) {
    gnrc_pktsnip_t *pkt;
    while(flags & THREAD_FLAG_RX_DONE) {
        if ((pkt = netif->ops->recv(netif))) {
           _process_receive_stats(netif, pkt);
            _pass_on_packet(pkt);
        }
        if (!tx_pkt) {
            tx_pkt = _send_queued_pkt(netif);
        }
    }
}

```

The `nrf24l01p_ng` has an internal queue.
Inside the `recv()` routine in the driver I would again call the netdev callback with `RX_DONE` to reset the thread flag.
Else, I guess a send call could override franes in the internal receive queue.

Reply:
This will not work so easy. The issue is that `thread_flags_wait_any(THREAD_FLAG_RX_DONE)` needs to be called in the loop. But that will block until the flag is set.

But I see no reason to extend the API to provide a non-blocking function that would to just allow that. But for now it might be the best if the driver would just return `-EBUSY` on TX if the receive queue is not yet drained.